flink-runtime/src/main/java/org/apache/flink/runtime/leaderretrieval/ZooKeeperLeaderRetrievalDriver.java

```diff
@@ -126,17 +161,13 @@ public void close() throws Exception {
 
         cache.close();
 
-        try {
-            if (client.getZookeeperClient().isConnected()
-                    && !connectionInformationPath.contains(RESOURCE_MANAGER_NODE)) {
-                client.watchers()
-                        .removeAll()
-                        .ofType(Watcher.WatcherType.Any)
-                        .forPath(connectionInformationPath);
-            }
-        } catch (KeeperException.NoWatcherException e) {
-            // Ignore the no watcher exception as it's just a safetynet to fix watcher leak issue.
-            // For more details, please refer to FLINK-33053.
+        if (client.getZookeeperClient().isConnected()
+                && watcherReferenceCounter.getAndIncrement() == 1) {
```

Comment:
getAndDecrement?

Reply:
yikes, I introduced that one when I did another code change. Good catch :+1: That proofs once more that a proper test should be added to cover this code path. :thinking: